File: 50dias/cpp/48.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <iomanip>
#include <list>
#include <ctime>
#include <conio.h>
#include <sstream>
#include <fstream>

using namespace std;

const int c_tamanhoProduto = 20;
const vector<int> codigos = { 100, 101, 102, 103, 104, 105 };
int codigoGeralPedido = 0;

map<int, float> precos = { {codigos[0], 1.20},
						   {codigos[1], 1.30},
						   {codigos[2], 1.50},
						   {codigos[3], 1.20},
						   {codigos[4], 1.30},
						   {codigos[5], 1.00} };

map<int, string> produtos = { {codigos[0], "Cachorro Quente"},
							  {codigos[1], "Bauru Simples"},
							  {codigos[2], "Bauru com ovo"},
							  {codigos[3], "Hamburguer"},
							  {codigos[4], "Cheesburguer"},
							  {codigos[5], "Refrigerante"} };

class Pedido {
public:
	Pedido operator+(Pedido& outro_pedido) {
		Pedido novoPedido;
		for (auto item : m_pedido) {
			int codigo = item.first;
			novoPedido.m_pedido[codigo] = m_pedido[codigo] + outro_pedido.m_pedido[codigo];
		}
		return novoPedido;
	}

	string CriaTextoPedido() const {
		stringstream stringPedido;
		stringPedido << "==============================================" << endl;
		stringPedido << "                   Pedido                     " << endl;
		stringPedido << "==============================================" << endl;
		stringPedido << "  Produto             |  Quant  |  Preco" << endl;
		stringPedido << "----------------------------------------------" << endl;
		float precoTotal = 0;
		for (auto item : m_pedido) {
			int codigo = item.first;
			int quantidade = item.second;
			if (quantidade > 0) {
				string nome = produtos[codigo];
				float preco = precos[codigo];
				stringPedido << "  " << nome << setw(c_tamanhoProduto - nome.size()) << " ";
				stringPedido << "|  " << setw(3) << quantidade << "    |  R$ ";
				stringPedido << std::fixed << std::setprecision(2);
				stringPedido << quantidade * preco << endl;
				precoTotal += quantidade * preco;
			}
		}
		stringPedido << "----------------------------------------------" << endl;
		stringPedido << "  Preco Total: R$ " << precoTotal << endl;
		stringPedido << "----------------------------------------------" << endl;
		return stringPedido.str();
	}

	void ImprimePedido() const {
		cout << CriaTextoPedido();
	}

protected:
	map<int, int> m_pedido = { {codigos[0], 0},
							   {codigos[1], 0},
							   {codigos[2], 0},
							   {codigos[3], 0},
							   {codigos[4], 0},
							   {codigos[5], 0} };
};

class PedidoUsuario : public Pedido {
public:
	PedidoUsuario(string nomeUsuario, int codigoPedido) :
		Pedido(),
		m_nomeUsuario(nomeUsuario),
		m_codigoPedido(codigoPedido)
	{
		m_data = time(0);
	}

	void RegistraPedido(int codigo, int quantidade) {
		m_pedido[codigo] += quantidade;
	}

	string CriaTextoRelatorio() const {
		stringstream stringRelatorio;

		char* dataHora = ctime(&m_data);

		stringRelatorio << endl;
		stringRelatorio << " Pedido n: " << m_codigoPedido << endl;
		stringRelatorio << " Cliente: " << m_nomeUsuario << endl;
		stringRelatorio << " Data: " << dataHora << endl;
		stringRelatorio << CriaTextoPedido();
		return stringRelatorio.str();

	}

	void ImprimeRelatorioPedido() const {
		cout << CriaTextoRelatorio();
	}

private:
	int m_codigoPedido;
	string m_nomeUsuario;
	time_t m_data;

};

list<PedidoUsuario> pedidosTotais;

void EscreveTextoNoArquivo(
	const string& nomeDoArquivo,
	const list<string>& texto)
{
	ofstream outputFile;
	outputFile.open(nomeDoArquivo, std::ios_base::app);

	for (auto linha : texto)
		outputFile << linha << endl;

	outputFile.close();
}

void ImprimeCardapio() {
	cout << "----------------------------------------------" << endl;
	cout << "  Produto             |  Codigo  |  Preco" << endl;
	cout << "----------------------------------------------" << endl;
	for (auto codigo : codigos) {
		string nome = produtos[codigo];
		float preco = precos[codigo];
		cout << "  " << nome << setw(c_tamanhoProduto - nome.size()) << " ";
		cout << "|  " << codigo << "     |  R$ ";
		cout << std::fixed << std::setprecision(2);
		cout << preco << endl;
	}
	cout << "----------------------------------------------" << endl;
}


void ImprimeMenu(const PedidoUsuario& pedidoAtual) {
	ImprimeCardapio();
	pedidoAtual.ImprimePedido();
}

void LeOpcoesDoUsuario(int& opcao, int& quantidade) {
	cout << "Digite o codigo do produto (0 para sair): ";
	cin >> opcao;
	quantidade = -1;
	if (opcao != 0) {
		cout << "Digite a quantidade: ";
		cin >> quantidade;
	}
}

bool ValidaInformacoes(int opcao, int quantidade) {
	bool produtoExiste = produtos.find(opcao) != produtos.end();
	bool quantidadeValida = quantidade > 0;
	if (!produtoExiste) {
		cout << endl << "  Codigo de produto nao encontrado!";
		cout << endl << endl;
	}
	else {
		if (!quantidadeValida) {
			cout << endl << "  Quantidade digitada eh invalida!";
			cout << endl << endl;
		}
	}

	return produtoExiste && quantidadeValida;
}

void LePedidoDoUsuario() {
	codigoGeralPedido++;
	string nomeUsuario;
	cout << "Digite o seu nome: ";
	getline(cin, nomeUsuario);

	PedidoUsuario pedidoAtual(nomeUsuario, codigoGeralPedido);

	int opcao, quantidade;
	system("cls");
	do {
		ImprimeMenu(pedidoAtual);

		LeOpcoesDoUsuario(opcao, quantidade);

		if (opcao != 0) {
			system("cls");
			if (ValidaInformacoes(opcao, quantidade)) {
				pedidoAtual.RegistraPedido(opcao, quantidade);
			}
		}

	} while (opcao != 0);

	pedidosTotais.push_back(pedidoAtual);
	cout << "\n  O seu pedido e o numero:" << codigoGeralPedido << endl;
	cout << "  Efetue o pagamento no caixa." << endl;
	cout << "  Muito obrigado pelo seu pedido!" << endl;
	cout << "\nTecle ENTER para sair!";
	getch();

}

int ImprimeMenuSistema() {
	cout << "==========================================" << endl;
	cout << "        Lancheria do Juquinha             " << endl;
	cout << "==========================================" << endl;
	cout << "  1 - Fazer pedido " << endl;
	cout << "  9 - Sair " << endl;
	cout << "------------------------------------------" << endl;
	int opcaoSistema;
	do {
		cout << "Digite a opcao: ";
		cin >> opcaoSistema;
		cin.ignore();
	} while (opcaoSistema != 1 and opcaoSistema != 9);
	return opcaoSistema;
}

string GeraCabecalhoRelatorio(string tituloCabecalho) {
	time_t agora = time(0);
	char* dataHora = ctime(&agora);
	stringstream cabecalhoRelatorio;
	cabecalhoRelatorio << "==============================================" << endl;
	cabecalhoRelatorio << "         " << tituloCabecalho << "          " << endl;
	cabecalhoRelatorio << "   " << dataHora << endl;
	cabecalhoRelatorio << "==============================================" << endl;
	return cabecalhoRelatorio.str();
}

void ImprimePedidosDoDia() {
	list<string> textoRelatorioPedidosDoDia;
	textoRelatorioPedidosDoDia.push_back(GeraCabecalhoRelatorio("RELATORIO DE PEDIDOS DO DIA"));
	for (auto pedido : pedidosTotais) {
		textoRelatorioPedidosDoDia.push_back(pedido.CriaTextoRelatorio());
	}

	EscreveTextoNoArquivo("pedidos.txt", textoRelatorioPedidosDoDia);
}

void ImprimeRelatorioVendasDia() {
	Pedido todosPedidosDoDia;
	for (auto pedido : pedidosTotais) {
		todosPedidosDoDia = todosPedidosDoDia + pedido;
	}

	list<string> textoRelatorioVendasDoDia;
	textoRelatorioVendasDoDia.push_back(GeraCabecalhoRelatorio("RELATORIO VENDAS DO DIA"));
	textoRelatorioVendasDoDia.push_back(todosPedidosDoDia.CriaTextoPedido());
	EscreveTextoNoArquivo("relatorio.txt", textoRelatorioVendasDoDia);
}

void ImprimeRelatoriosDoDia() {
	ImprimePedidosDoDia();
	ImprimeRelatorioVendasDia();
}

int main()
{
	bool continuarSistema = false;
	int quantidade = 0;
	do {
		system("cls");
		int opcaoSistema = ImprimeMenuSistema();
		continuarSistema = opcaoSistema != 9;
		if (continuarSistema) {
			LePedidoDoUsuario();
		}
	} while (continuarSistema);

	ImprimeRelatoriosDoDia();
	return 0;
}

File: 50dias/python/48.py
import os
from datetime import datetime

prodStringSize = 20
codigos = [100, 101, 102, 103, 104, 105]
produtos = {codigos[0]: {'nome': 'Cachorro Quente',
                         'preco': 1.20},
            codigos[1]: {'nome': 'Bauru Simples',
                         'preco': 1.30},
            codigos[2]: {'nome': 'Bauru com ovo',
                         'preco': 1.50},
            codigos[3]: {'nome': 'Hamburguer',
                         'preco': 1.20},
            codigos[4]: {'nome': 'Cheeseburguer',
                         'preco': 1.30},
            codigos[5]: {'nome': 'Refrigerante',
                         'preco': 1.00}}

pedidosTotais = {}
codigoPedido = 0


def EscreveTextoNoArquivo(texto, nomeDoArquivo):
    output_file = open(nomeDoArquivo, 'a')
    output_file.write(texto)
    output_file.close()

def ImprimeCardapio():
    print('------------------------------------------')
    print('  Produto             |  Codigo  |  Preco ')
    print('------------------------------------------')
    for codigo in codigos:
        produto = produtos[codigo]
        nome = produto['nome'].ljust(prodStringSize)
        preco = produto['preco']
        msg = '  {nome}|  {cod}     |  R$ {preco:.2f}'.format(nome=nome,
                                                              cod=codigo,
                                                              preco=preco)
        print(msg)
    print('------------------------------------------')


def GeraTextoPedido(pedidoAtual):
    precoTotal = 0.00
    textoPedido = '==========================================\n'
    textoPedido += '                 Pedido                 \n'
    textoPedido +='==========================================\n'
    textoPedido +='  Produto             |  Quant   |  Preco \n'
    textoPedido +='------------------------------------------\n'
    for codigo in codigos:
        quantidade = pedidoAtual[codigo]
        if quantidade > 0:
            produto = produtos[codigo]
            nome = produto['nome'].ljust(prodStringSize)
            preco = produto['preco']
            msg = '  {nome}|  {quant}     |  R$ {preco:.2f}\n'.format(nome=nome,
                                                                    quant=quantidade,
                                                                    preco=quantidade * preco)
            textoPedido += msg
            precoTotal += quantidade * preco

    textoPedido += '------------------------------------------\n'
    textoPedido += '  Preco Total: R$ {preco:.2f}\n'.format(preco=precoTotal)
    textoPedido += '------------------------------------------\n'
    return textoPedido

def ImprimePedido(pedidoAtual):
    textoPedido = GeraTextoPedido(pedidoAtual)
    print(textoPedido)

def ImprimeMenuPedidos(pedidoAtual):
    ImprimeCardapio()
    ImprimePedido(pedidoAtual)

def RegistraPedido(pedidoAtual, codigo, quantidade):
    pedidoAtual[codigo] += quantidade


def LeOpcoesDoUsuario():
    opcao = int(input('Digite o codigo do produto (0 para sair): '))
    quantidade = -1
    if opcao != 0:
        quantidade = int(input('Digite a quantidade: '))
    return opcao, quantidade


def ValidaInformacoes(opcao, quantidade):
    produtoExiste = opcao in codigos
    quantidadeValida = quantidade > 0
    if not produtoExiste:
        print('\n  Codigo de produto nao encontrado!\n')
    elif not quantidadeValida:
        print('\n  Quantidade digitada eh invalida!\n')

    return produtoExiste and quantidadeValida


def CriaPedido():
    pedido = {codigos[0]: 0,
              codigos[1]: 0,
              codigos[2]: 0,
              codigos[3]: 0,
              codigos[4]: 0,
              codigos[5]: 0}
    return pedido


def LePedidoDoUsuario():
    global codigoPedido
    codigoPedido += 1
    nomeUsuario = input('Digite o seu nome: ')
    pedidoAtual = CriaPedido()
    opcao = -1
    os.system('cls')
    while opcao != 0:
        ImprimeMenuPedidos(pedidoAtual)
        opcao, quantidade = LeOpcoesDoUsuario()
        if opcao != 0:
            os.system('cls')
            if ValidaInformacoes(opcao, quantidade):
                RegistraPedido(pedidoAtual, opcao, quantidade)

    agora = datetime.now()
    data = agora.strftime("%d/%m/%Y %H:%M:%S")
    pedidosTotais[codigoPedido] = {'nome': nomeUsuario,
                                   'data': data,
                                   'pedido': pedidoAtual}

    print('\n  O seu pedido é o número: {}'.format(codigoPedido))
    print('  Efetue o pagamento no caixa.')
    print('  Muito obrigado pelo seu pedido!')
    _ = input('\nTecle ENTER para sair!')


def ImprimeMenuSistema():
    print('==========================================')
    print('        Lancheria do Juquinha             ')
    print('==========================================')
    print('  1 - Fazer pedido ')
    print('  9 - Sair ')
    print('------------------------------------------')
    opcao = -1
    while opcao != 1 and opcao != 9:
        opcao = int(input('Digite a opcao: '))
    return opcao

def GeraTextoPedidosDoDia():
    textoPedidosDia = ""
    for codigoPedido, pedidoUsuario in pedidosTotais.items():
        textoPedidosDia += ' Pedido n: {}\n'.format(codigoPedido)
        textoPedidosDia += ' Cliente: {}\n'.format(pedidoUsuario['nome'])
        textoPedidosDia += ' Data: {}\n'.format(pedidoUsuario['data'])
        textoPedidosDia += GeraTextoPedido(pedidoUsuario['pedido'])
    return textoPedidosDia

def GeraCabecalhoRelatorio(tituloCabecalho):
    agora = datetime.now()
    data = agora.strftime("%d/%m/%Y %H:%M:%S")

    textoCabecalho = "==============================================\n"
    textoCabecalho += "          " + tituloCabecalho + "\n"
    textoCabecalho += "               " + data + "\n"
    textoCabecalho += "==============================================\n"
    return textoCabecalho

def ImprimePedidosDoDia():
    textoRelatorioPedidosDoDia = GeraCabecalhoRelatorio("RELATORIO DE PEDIDOS DO DIA")
    textoRelatorioPedidosDoDia += GeraTextoPedidosDoDia()
    EscreveTextoNoArquivo(textoRelatorioPedidosDoDia, 'pedidos.txt')

def ImprimeRelatorioVendasDia():
    todosPedidosDoDia = CriaPedido()
    for codigoPedido, pedidoUsuario in pedidosTotais.items():
        for codigoProduto, quantidade in pedidoUsuario['pedido'].items():
            RegistraPedido(todosPedidosDoDia,
                           codigoProduto,
                           quantidade)

    textoRelatorioVendas = GeraCabecalhoRelatorio("RELATORIO VENDAS DO DIA")
    textoRelatorioVendas += GeraTextoPedido(todosPedidosDoDia)
    EscreveTextoNoArquivo(textoRelatorioVendas, 'relatorio.txt')

def ImprimeRelatoriosDoDia():
    ImprimePedidosDoDia()
    ImprimeRelatorioVendasDia()

def main():
    opcaoSistema = -1
    while opcaoSistema != 9:
        os.system('cls')
        opcaoSistema = ImprimeMenuSistema()
        if opcaoSistema != 9:
            LePedidoDoUsuario()

    ImprimeRelatoriosDoDia()


main()
